datasetinsights/io/tracker/factory.py

```diff
@@ -0,0 +1,115 @@
+    CLIENT_ID = "client_id"
+    HOST_ID = "host"
+    EXP_NAME = "experiment"
+    RUN_NAME = "run"
+    MLFLOW_TRACKER = "mlflow"
+    DEFAULT_RUN_NAME = "run_" + TIMESTAMP_SUFFIX
+    __singleton_lock = threading.Lock()
+    __tracker_instance = None
+    RUN_FAILED = "FAILED"
+
+    @staticmethod
+    def create(config, tracker_type):
+        """create tracker class object.
+
+        Args:
+            config : config object, holds server details
+            tracker_type : type of tracker
+
+        Returns:
+            tracker instance.
+        """
+        if TrackerFactory.MLFLOW_TRACKER == tracker_type:
+
+            tracker = config.get(TrackerFactory.TRACKER, None)
+            if tracker and tracker.get(TrackerFactory.MLFLOW_TRACKER, None):
+                mlflow_config = tracker.get(TrackerFactory.MLFLOW_TRACKER)
+                host_id = mlflow_config.get(TrackerFactory.HOST_ID, None)
+                if host_id:
+                    client_id = mlflow_config.get(
+                        TrackerFactory.CLIENT_ID, None
+                    )
+                    exp_name = mlflow_config.get(TrackerFactory.EXP_NAME, None)
+                    TrackerFactory.update_run_name(mlflow_config)
+                    mlflow_tracker = TrackerFactory._mlflow_tracker_instance(
+                        host_id=host_id, client_id=client_id, exp_name=exp_name
+                    ).get_mlflow()
+                    mlflow_tracker.start_run(
+                        run_name=TrackerFactory.DEFAULT_RUN_NAME
+                    )
+                    return mlflow_tracker
+            return TrackerFactory._null_tracker()
+        else:
+            raise NotImplementedError(f"Unknown tracker {tracker_type}!")
+
+    @staticmethod
+    def _mlflow_tracker_instance(host_id=None, client_id=None, exp_name=None):
+
+        """Static instance access method.
+
+        Args:
+            host_id: MlTracker server host
+            client_id: MLFlow tracking server client id
+            exp_name: name of the experiment
+        Returns:
+            tracker singleton instance.
+        """
+        if not TrackerFactory.__tracker_instance:
+            with TrackerFactory.__singleton_lock:
+                if not TrackerFactory.__tracker_instance:
+                    TrackerFactory.__tracker_instance = MLFlowTracker(
+                        host_id=host_id, client_id=client_id, exp_name=exp_name
+                    )
+        return TrackerFactory.__tracker_instance
+
+    @staticmethod
+    def _null_tracker():
+        """private getter method to get singleton instance.
+
+        Returns:
+            NullTracker singleton instance.
+        """
+        if not TrackerFactory.__tracker_instance:
+            with TrackerFactory.__singleton_lock:
+                if not TrackerFactory.__tracker_instance:
+                    TrackerFactory.__tracker_instance = NullTracker()
+        return TrackerFactory.__tracker_instance
```

Comment:
can't the null tracker be static since it doesn't need to keep any state?

Reply:
- What do you mean by static NullTracker? 
- We need to return some instance on which logging can be handle.

Reply:
Did you successful make reference to NullTracker class without creating an instance?

Reply:
With reference  __getattr__(self, name) does not get called.